### ISSUE TITLE
Programming ipac2 board occasionally miss programs.

### ISSUE THREAD

**Issue by OzFalcon** [katie-snow/Ultimarc-linux#25]
Noticed this on 2014 boards, Have not tested on 2015 boards yet.
Every now and then the key assignments get messed up, Keys are assigned to wrong pins.
Seems to happen every 10~15 programs.

Has happened before, But I just put it down to updated libraries or something.
But now I don't think so, Because I can reproduce it just by repeat programming.
It's frequency of failure is enough to raise an issue.

Seems similar to an issue rootrunner was having.
https://github.com/katie-snow/Ultimarc-linux/issues/9

**Comment by katie-snow** [comment]
I have had the same problem with the IPAC4 board and it seems like the frequency of failure has increased when I added the macro data.
Looking around for an alternative to libusb led me to hidraw library http://www.signal11.us/oss/hidapi/ (https://github.com/signal11/hidapi).  I choose the implementation that uses the kernels hidraw functionality to write to the board.  This still didn't fix the error, there may have been a reduction in failure frequency, but i wasn't checking very closely.

I need to look at libusb_open_device_with_vid_pid function and the device lookup functions from libusb to see if there is a difference between the propagation of the underlining structure.

**Comment by OzFalcon** [comment]
Yes, I was reading on libhid which is now obsolete and recommended to use hidapi.
While http://www.libusb.org/ is not obsolete, It's good to try alternatives.

I know very little (Nothing) about USB device programming, But is it possible to lower the connection speed of the USB device while doing the data transfer?

**Comment by OzFalcon** [comment]
Just looking at your code and I noticed 5 bytes for data block length.
ipacseries.h
#define IPACSERIES_MESG_LENGTH 5

Then you appear to be sending the data in 4 byte blocks.
(Unless I misunderstand how it's supposed to be written) 
ipacseries.c
while (pos < size)
  {
    memcpy(&mesg[1], &barray[pos], 4);

    debug ("Writing (%i): %x, %x, %x, %x", pos, mesg[1], mesg[2], mesg[3], mesg[4]);
    if (transfer)
    {
      ret = libusb_control_transfer(handle,
                                    UM_REQUEST_TYPE,
                                    UM_REQUEST,
                                    IPACSERIES_VALUE,
                                    interface,
                                    mesg,
                                    IPACSERIES_MESG_LENGTH,
                                    UM_TIMEOUT);
      debug ("Write result: %i", ret);
    }
    pos+=4;
  }

**Comment by OzFalcon** [comment]
Just to note, Travis's IPacUtil sends the data to the older cards in x3 32 byte blocks.

**Comment by molith-modragon** [comment]
The data is sent in 4 byte blocks, but a single USB transfer seems to need some sort of command byte (msg[0]). So the data sent is:
```C++
msg[0], msg[1], msg[2], msg[3], msg[4]
```
which corresponds to:
```C++
cmd, data1, data2, data3, data4
cmd, data5, data6, data7, data8
cmd, data9, data10, data11, data12
```
msg[0] is the command byte and can be found in the initialization of mesg in ipacseries.c writeIPACSeriesUSB, where it is set to 0x03.

Therefore you see a message length of 5 bytes, but only 4 data bytes per iteration.

**Comment by OzFalcon** [comment]
Now I see how 0x03 is used in  a 5 byte block, Thank you.
Do you mean some sort of delay between programming each block or each programming cycle?

**Comment by molith-modragon** [comment]
This depends on the specs of the used chips as well as the corresponding programming interface.

E.g. the PacLED64 uses LED driver chips that are controlled via i2c bus. This bus has a specific timing that is significantly slower than USB. If you send too many commands in this situation, the commands are faster than the i2c bus.

Of course, this is a very rough description, as the controller chip on the PacLED64 has some firmware that would handle the incoming commands and send them to the corresponding LED driver chips, and especially the script mode on the PacLED64 is driven by that controller chip.

Anyway, I would try the following:
- Find some spec about the used chips. Maybe there is some timing constraint.
- Add some small delay to each USB command.
- If there is some erasing sequence in front, add some delay afterwards.

**Comment by OzFalcon** [comment]
I have just done some testing on this issue and found it to be mainly CPU load related.
ie. The greater your CPU load (Background activity), The more chance of program failure.
  The problem appears to be considerably worse when "autoconnect" is enabled.

CPU Load:
To load my old dual core pc, I used the program "stress" with x2 work threads.
>stress --cpu 2

Fail example:
With CPU load & "autoconnect" enabled, Programming failure is ~10 cycles.

Recommend:
I highly recommend disabling all pre2015 ipacs "autoconnect" option.

Pending:
I will take a look at the post 2015 cards to see if the same problem exists.

**Comment by OzFalcon** [comment]
A quick look at the post 2015 cards shows the problem does not exist for these cards.
For 2015 cards "autoconnect" can be left enabled, and programming with CPU load is trouble free.

I will create a fork with "autoconnect" disabled for the pre 2015 cards.

**Comment by katie-snow** [comment]
Thanks for the help and suggestions.  I have a new little one in the house, so the changes will be a little bit slower coming.

**Comment by OzFalcon** [comment]
Congratulations, Keep up the good work and take your time :)

**Comment by OzFalcon** [comment]
Just to add to this issue,
This problem seems to occur more frequently when booting from a power off state.
eg. When I turn my cabinet on from the wall switch, 2014 ipac miss programs on about 20% of the boots.

**Comment by katie-snow** [comment]
OzFalcon,

I have something for you to try, if you are up for it.  I think the problem is permissions on the USB device location in the file system.  What I would like you to do is change the 21-ultimarc.rules file in /etc/udev/rules.d/.  Change your device entry so it looks like the following.
ACTION=="add", SUBSYSTEMS=="usb", ATTRS{idVendor}=="d208", ATTRS{idProduct}=="0310", GROUP:="ultimarc"

Remove MODE and add GROUP:="ultimarc"

Then you will need to create that group ultimarc and add your user to that group.  Then restart the computer and see how often it fails.  You can look at the path to make sure they have their group changed.  /dev/input/.

I have done this and I think am having better success.  Its not solving the problem completely but the failures seem more infrequent.  That is why I would like you to give this a try.

Thanks
Katie

**Comment by OzFalcon** [comment]
Katie,
Ok I'll set some time aside to do some investigation. Thanks for the heads up.
